source/common/router/router.cc

```diff
@@ -485,7 +486,21 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
   }
 
   // Fetch a connection pool for the upstream cluster.
-  Http::ConnectionPool::Instance* conn_pool = getConnPool();
+  Http::ConnectionPool::Instance* conn_pool;
+  auto url_str = headers.Host()->value().getStringView();
+  const auto parsed_authority = Http::Utility::parseAuthority(url_str.data());
+  const auto http_protocol_options = cluster_->httpProtocolOptions();
+
+  if (http_protocol_options != absl::nullopt && http_protocol_options->auto_sni() &&
+      !parsed_authority.is_ip_address_) {
+    callbacks_->streamInfo().filterState().setData(
+        Network::UpstreamServerName::key(), std::make_unique<Network::UpstreamServerName>(url_str),
+        StreamInfo::FilterState::StateType::Mutable);
```

Comment:
Can you add a comment about what this actually does with a refererence to the connection pool code which will use this option to determine whether to make a new connection or not. This is a bit magic.